File: tests/build-verify.js
#!/usr/bin/env node
/**
 * 千界花园 — 构建验证脚本 (Build Verification)
 * 验证前端代码完整性、类型检查、路由对齐
 * 
 * 用法:
 *   node tests/build-verify.js
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

const C = {
  reset: '\x1b[0m', green: '\x1b[32m', red: '\x1b[31m',
  yellow: '\x1b[33m', cyan: '\x1b[36m', dim: '\x1b[2m',
};

function ok(m) { console.log(`${C.green}✓${C.reset} ${m}`); }
function fail(m) { console.log(`${C.red}✗${C.reset} ${m}`); }
function info(m) { console.log(`${C.cyan}ℹ${C.reset} ${m}`); }

const ROOT = path.resolve(__dirname, '..');
const FRONTEND = path.join(ROOT, 'frontend');
const BACKEND = path.join(ROOT, 'backend');

const results = [];

function check(name, fn) {
  try {
    fn();
    ok(name);
    results.push({ name, status: 'PASS' });
  } catch (err) {
    fail(`${name} — ${err.message}`);
    results.push({ name, status: 'FAIL', error: err.message });
  }
}

console.log(`\n${C.cyan}╔══════════════════════════════════════════════════════════════╗${C.reset}`);
console.log(`${C.cyan}║      千界花园 — 构建验证脚本 v1.0                           ║${C.reset}`);
console.log(`${C.cyan}╚══════════════════════════════════════════════════════════════╝${C.reset}\n`);

// ── 1. 目录结构完整性 ──────────────────────────────────────────
check('frontend/src/pages 存在且包含 >=50 个页面', () => {
  const dir = path.join(FRONTEND, 'src/pages');
  if (!fs.existsSync(dir)) throw new Error('目录不存在');
  const files = fs.readdirSync(dir).filter(f => f.endsWith('.tsx'));
  if (files.length < 50) throw new Error(`仅 ${files.length} 个页面`);
});

check('frontend/src/components 存在', () => {
  const dir = path.join(FRONTEND, 'src/components');
  if (!fs.existsSync(dir)) throw new Error('目录不存在');
});

check('backend/src/routes 存在且 >=25 个路由', () => {
  const dir = path.join(BACKEND, 'src/routes');
  if (!fs.existsSync(dir)) throw new Error('目录不存在');
  const files = fs.readdirSync(dir).filter(f => f.endsWith('.ts'));
  if (files.length < 25) throw new Error(`仅 ${files.length} 个路由`);
});

check('backend/src/services 存在且 >=20 个服务', () => {
  const dir = path.join(BACKEND, 'src/services');
  if (!fs.existsSync(dir)) throw new Error('目录不存在');
  const files = fs.readdirSync(dir).filter(f => f.endsWith('.ts'));
  if (files.length < 20) throw new Error(`仅 ${files.length} 个服务`);
});

check('backend/src/adapters 存在', () => {
  const dir = path.join(BACKEND, 'src/adapters');
  if (!fs.existsSync(dir)) throw new Error('目录不存在');
});

// ── 2. App.tsx 路由完整性 ─────────────────────────────────────
check('App.tsx 包含 >=50 条 Route', () => {
  const appPath = path.join(FRONTEND, 'src/App.tsx');
  const content = fs.readFileSync(appPath, 'utf-8');
  const matches = content.match(/<Route\s+path=/g);
  const count = matches ? matches.length : 0;
  if (count < 50) throw new Error(`仅 ${count} 条路由`);
});

check('App.tsx 所有 lazy import 都有对应文件', () => {
  const appPath = path.join(FRONTEND, 'src/App.tsx');
  const content = fs.readFileSync(appPath, 'utf-8');
  const imports = [...content.matchAll(/lazy\(\(\) => import\('@\/pages\/(\w+)'\)/g)];
  const pagesDir = path.join(FRONTEND, 'src/pages');
  for (const [, name] of imports) {
    const filePath = path.join(pagesDir, `${name}.tsx`);
    if (!fs.existsSync(filePath)) throw new Error(`缺少页面: ${name}.tsx`);
  }
});

// ── 3. 关键配置文件存在 ──────────────────────────────────────
check('docker-compose.yml 存在且包含 >=3 个服务', () => {
  const fp = path.join(ROOT, 'docker-compose.yml');
  if (!fs.existsSync(fp)) throw new Error('文件不存在');
  const content = fs.readFileSync(fp, 'utf-8');
  const services = content.match(/^\s{2}[\w-]+:/gm);
  const count = services ? services.length : 0;
  if (count < 3) throw new Error(`仅 ${count} 个服务定义`);
});

check('openapi.yaml 存在且有效', () => {
  const fp = path.join(ROOT, 'openapi.yaml');
  if (!fs.existsSync(fp)) throw new Error('文件不存在');
  const content = fs.readFileSync(fp, 'utf-8');
  if (!content.includes('openapi:')) throw new Error('非有效 OpenAPI 格式');
});

check('package.json 根项目存在', () => {
  const fp = path.join(ROOT, 'package.json');
  if (!fs.existsSync(fp)) throw new Error('文件不存在');
});

check('frontend/package.json 存在', () => {
  const fp = path.join(FRONTEND, 'package.json');
  if (!fs.existsSync(fp)) throw new Error('文件不存在');
});

check('backend/package.json 存在', () => {
  const fp = path.join(BACKEND, 'package.json');
  if (!fs.existsSync(fp)) throw new Error('文件不存在');
});

// ── 4. Electron 打包配置 ─────────────────────────────────────
check('electron/main.js 存在', () => {
  const fp = path.join(ROOT, 'electron/main.js');
  if (!fs.existsSync(fp)) throw new Error('文件不存在');
});

check('electron/preload.js 存在', () => {
  const fp = path.join(ROOT, 'electron/preload.js');
  if (!fs.existsSync(fp)) throw new Error('文件不存在');
});

// ── 5. 测试脚本完整性 ──────────────────────────────────────────
check('test_all_providers.py 存在且包含10个Provider', () => {
  const fp = path.join(ROOT, 'test_all_providers.py');
  if (!fs.existsSync(fp)) throw new Error('文件不存在');
  const content = fs.readFileSync(fp, 'utf-8');
  const providers = ['kimi-code', 'openai', 'anthropic', 'deepseek', 'gemini', 'qwen', 'glm', 'openrouter', 'moonshot', 'azure-openai'];
  for (const p of providers) {
    if (!content.includes(p)) throw new Error(`缺少 ${p}`);
  }
});

check('test_kimi_layer1.py 存在', () => {
  const fp = path.join(ROOT, 'test_kimi_layer1.py');
  if (!fs.existsSync(fp)) throw new Error('文件不存在');
});

check('BUILD_GUIDE.md 存在', () => {
  const fp = path.join(ROOT, 'BUILD_GUIDE.md');
  if (!fs.existsSync(fp)) throw new Error('文件不存在');
});

// ── 6. 前端关键页面内容检查 ────────────────────────────────────
check('APIKeys.tsx 包含 Provider 选择器', () => {
  const fp = path.join(FRONTEND, 'src/pages/APIKeys.tsx');
  if (!fs.existsSync(fp)) throw new Error('文件不存在');
  const content = fs.readFileSync(fp, 'utf-8');
  if (!content.includes('provider') && !content.includes('Provider')) {
    throw new Error('缺少 Provider 相关逻辑');
  }
});

check('UploadsPage.tsx 包含知识库关联', () => {
  const fp = path.join(FRONTEND, 'src/pages/UploadsPage.tsx');
  if (!fs.existsSync(fp)) throw new Error('文件不存在');
  const content = fs.readFileSync(fp, 'utf-8');
  if (!content.includes('knowledge') && !content.includes('knowledgeBase')) {
    throw new Error('缺少知识库关联逻辑');
  }
});

check('Chat.tsx 使用对话兼容层', () => {
  const fp = path.join(FRONTEND, 'src/pages/Chat.tsx');
  if (!fs.existsSync(fp)) throw new Error('文件不存在');
  const content = fs.readFileSync(fp, 'utf-8');
  if (!content.includes('sendMessage') && !content.includes('message')) {
    throw new Error('缺少对话相关逻辑');
  }
});

// ── 7. 后端关键服务检查 ────────────────────────────────────────
check('UnifiedLLMAdapter.ts 存在', () => {
  const fp = path.join(BACKEND, 'src/services/UnifiedLLMAdapter.ts');
  if (!fs.existsSync(fp)) throw new Error('文件不存在');
});

check('LLMProviderRegistry.ts 存在', () => {
  const fp = path.join(BACKEND, 'src/services/LLMProviderRegistry.ts');
  if (!fs.existsSync(fp)) throw new Error('文件不存在');
});

check('IntentClassifier.ts 存在且包含 skills', () => {
  const fp = path.join(BACKEND, 'src/services/IntentClassifier.ts');
  if (!fs.existsSync(fp)) throw new Error('文件不存在');
  const content = fs.readFileSync(fp, 'utf-8');
  if (!content.includes('skill') && !content.includes('intent')) {
    throw new Error('缺少技能/意图相关逻辑');
  }
});

check('KimiClusterOrchestrator.ts 存在', () => {
  const fp = path.join(BACKEND, 'src/KimiClusterOrchestrator.ts');
  if (!fs.existsSync(fp)) throw new Error('文件不存在');
});

// ── 8. TypeScript 类型检查 (如可用) ────────────────────────────
try {
  const tscPath = path.join(FRONTEND, 'node_modules/.bin/tsc');
  if (fs.existsSync(tscPath)) {
    check('frontend TypeScript 类型检查通过', () => {
      try {
        execSync(`${tscPath} --noEmit`, { cwd: FRONTEND, stdio: 'pipe', timeout: 60000 });
      } catch (e) {
        const out = e.stdout?.toString() || '';
        const err = e.stderr?.toString() || '';
        // 允许存在 MUI/Emotion 警告，但不得有致命错误
        const fatal = (out + err).split('\n').filter(l => l.includes('error TS') && !l.includes('TS7016') && !l.includes('TS2307'));
        if (fatal.length > 0) throw new Error(`致命类型错误: ${fatal.length} 个`);
      }
    });
  } else {
    info('跳过 TypeScript 类型检查 (未安装依赖)');
  }
} catch {
  info('跳过 TypeScript 类型检查 (tsc 不可用)');
}

// ── 汇总 ───────────────────────────────────────────────────────
const pass = results.filter(r => r.status === 'PASS').length;
const failCount = results.filter(r => r.status === 'FAIL').length;
const total = results.length;

console.log(`\n${C.cyan}──────────────────────────────────────────────────────────────${C.reset}`);
console.log(`  总计: ${total}  |  ✅ 通过: ${pass}  |  ❌ 失败: ${failCount}  |  通过率: ${Math.round(pass/total*100)}%`);
console.log(`${C.cyan}──────────────────────────────────────────────────────────────${C.reset}\n`);

process.exit(failCount > 0 ? 1 : 0);

File: tests/python/build_verify.py
#!/usr/bin/env python3
"""
千界花园 - 前端TypeScript类型检查与路由覆盖率验证
验证: 1) 前端无TypeScript错误 2) 每个后端路由都有对应前端页面
用法: python build_verify.py [--skip-tsc]
"""

import subprocess, sys, os, json, re
from pathlib import Path
from collections import defaultdict

def run_tsc(frontend_dir: Path):
    print("🔍 正在运行 TypeScript 类型检查...")
    print(f"   目录: {frontend_dir}")
    try:
        result = subprocess.run(
            ["npx", "tsc", "--noEmit", "--pretty", "false"],
            capture_output=True, text=True, timeout=300, cwd=str(frontend_dir)
        )
        errors = [l for l in result.stdout.split("\n") if "error TS" in l]
        if not errors and result.returncode == 0:
            print("   ✅ 0 个TypeScript错误")
            return True, 0, ""
        else:
            # 统计
            unique_files = set()
            for e in errors:
                m = re.match(r"([^()]+)\(\d+,\d+\)", e)
                if m:
                    unique_files.add(m.group(1).strip())
            print(f"   ❌ {len(errors)} 个错误, 涉及 {len(unique_files)} 个文件")
            return False, len(errors), result.stdout
    except subprocess.TimeoutExpired:
        print("   ⏰ TypeScript检查超时")
        return False, -1, "timeout"
    except FileNotFoundError:
        print("   ⚠️ 未找到npx/tsc，跳过类型检查（需本地运行: npm install && npx tsc --noEmit）")
        return True, 0, "skipped"

def check_route_coverage(root: Path):
    print("\n🔍 检查前后端路由覆盖率...")
    backend_routes = root / "backend" / "src" / "routes"
    frontend_pages = root / "frontend" / "src" / "pages"
    app_tsx = root / "frontend" / "src" / "App.tsx"

    # 提取后端路由名称
    be_routes = set()
    if backend_routes.exists():
        for f in backend_routes.glob("*.ts"):
            name = f.stem.replace("-", "").lower()
            be_routes.add(name)

    # 提取前端页面名称
    fe_pages = set()
    page_map = {}
    if frontend_pages.exists():
        for f in frontend_pages.glob("*.tsx"):
            name = f.stem.lower()
            fe_pages.add(name)
            page_map[name] = f.name

    # 从App.tsx提取路由
    app_routes = set()
    if app_tsx.exists():
        content = app_tsx.read_text(encoding="utf-8")
        for line in content.split("\n"):
            if "path=" in line and '"/api/' not in line:
                m = re.search(r'path=["\']([^"\']+)["\']', line)
                if m:
                    route = m.group(1).strip("/").lower().replace("-", "").replace("/", "")
                    app_routes.add(route)

    print(f"   后端路由文件: {len(be_routes)} 个")
    print(f"   前端页面组件: {len(fe_pages)} 个")
    print(f"   App.tsx路由: {len(app_routes)} 条")

    # 尝试匹配
    matched = 0
    unmatched = []
    for be in sorted(be_routes):
        # 尝试各种命名变体匹配
        variants = [be, be.replace("s", ""), be + "s", be.replace("center", ""), be.replace("manager", ""), be.replace("monitor", ""), be + "page"]
        found = any(v in fe_pages or v in app_routes for v in variants)
        if found:
            matched += 1
        else:
            unmatched.append(be)

    print(f"\n   ✅ 匹配页面: {matched}/{len(be_routes)}")
    if unmatched:
        print(f"   ⚠️ 可能缺失前端页面: {', '.join(unmatched[:10])}")
    else:
        print("   🎉 所有后端路由都有对应前端页面！")

    return matched, len(be_routes), unmatched

def main():
    root = Path(__file__).parent.parent.parent
    frontend_dir = root / "frontend"

    skip_tsc = "--skip-tsc" in sys.argv

    ok = True
    if not skip_tsc:
        tsc_ok, err_count, tsc_out = run_tsc(frontend_dir)
        ok = ok and tsc_ok
        if not tsc_ok and tsc_out and tsc_out != "skipped":
            print("\n--- TypeScript 错误详情 (前20条) ---")
            for line in tsc_out.split("\n")[:20]:
                if line.strip():
                    print(line)
    else:
        print("⏭️ 跳过TypeScript检查")

    matched, total, unmatched = check_route_coverage(root)
    coverage_ok = matched >= total * 0.9

    print(f"\n{'='*60}")
    if ok and coverage_ok:
        print("🟢 构建验证通过")
        sys.exit(0)
    else:
        print("🟡 构建验证有警告/错误")
        sys.exit(1)

if __name__ == "__main__":
    main()
